Custom keyboard shortcuts

MichaelBuessemeyer opened the pull request:
Architecture design doc: https://www.notion.so/scalableminds/Design-Doc-Configurable-Keyboard-Shortcuts-Architecture-34bb51644c63806e9717f96af0aa62b9#34bb51644c63807eb167ea6563284892

### URL of deployed dev instance (used for testing):
- https://customkeyboardshortcuts.webknossos.xyz/

### Steps to test:
- Try out the existing shortcuts
- Should be working as before
- open the shortcuts modal
  - mouse shortcut should be listed
  - adding new keyboard shortcuts, removing existing ones, saving, persistence, restoring defaults should work
  - check whether tool-specific shortcuts work properly
  - check whether shortcut collisions are detected properly
  - check that shortcut recording works
  - Add shortcuts using special symbols on your keyboard like + or -
    - Should directly be displayed correctly in Edge / Chrome
    - In FF the mentioned build up of the layout map should work
 
-  Test layout map in FF:
  - The shortcut modal should have a warning text about eventual mismatch of the displayed keys and the acutal keys on the keyboard. 
  - Configure a shortcut with `+` on a german layout
  - should directly show up correctly during recording and when added to the modal
  - hit save and check whether the shortcut works by hitting + -> should work
  - reload the page and check the shortcut again -> should still work
  - open the shortcut modal -> the shortcut should be displayed as `+` and not as the US equivalent
  - clear the local storage and reload
  - !FIRST! open the shortcut modal -> the `+` shortcut should now be displayed as if you had a US keyboard layout.
  - Press the `+` button -> now the shortcut should be displayed correctly
  - test the shortcut, it should work
  - switch the layout -> the shortcut should still work
  - open the shortcut modal again -> should show the shortcut of the current layout. If other shortcuts are configured with special symbols, they might not be displayed correctly due to the layout change but should be displayed correctly once pressed.    
  
  
### TODOs:
- [x] Use Keystrokes
- [x] ... ? Migrate all keyboard shortcuts
- [x] Make it compile :D
- [x] Do live collision detection during recording shortcuts
- [x] List all mouse shortcuts but not editable -> direct docs in WK!
  - continue at: https://docs.webknossos.org/webknossos/ui/keyboard_shortcuts.html#mesh-related-shortcuts
- [x] Persist -> maybe in userConfig in store / backend? Or a separate field?
  - [x] Do not send update to backend if not logged in 
- [x] Adjust suggested shortcuts in info tab
  - [x] implement
  - [x] check if correct
- [x] Fix new "mouse move" icon
> <img width="486" height="238" alt="image" src="https://github.com/user-attachments/assets/dda399b4-559a-47b1-9f55-85a70ff39186" />
- [x] re-enable TODO comment check
- [x] Recorder modal allows multi setting the same shortcut
- [x] Recorder modal: Editing  one shortcut marks it initally as a collision


### Issues:
- fixes #8175

------
(Please delete unneeded items, merge only when none are left open)
- [x] Added changelog entry (create a `$PR_NUMBER.md` file in `unreleased_changes` or use `./tools/create-changelog-entry.py`)
- [x] Needs sql migration 
- [x] Updated [documentation](../blob/master/docs) if applicable
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md) -> FF and layout switching are kinda like edge cases

MichaelBuessemeyer committed:
Add initial working version fro controller shortcuts

MichaelBuessemeyer committed:
WIP add arbitary controller shortcuts support

MichaelBuessemeyer committed:
WIP add arbitrary controller support

MichaelBuessemeyer committed:
refactor key config format to support multiple shortcuts per handler and WIP add arbitrary controller shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
sort imports

MichaelBuessemeyer committed:
WIP: add plane controller shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix keyboard_shortcut_constants.ts

MichaelBuessemeyer committed:
WIP migrate plane shortcuts & tool specific shortcuts to new "shortcut creation mechanism"

MichaelBuessemeyer committed:
refactor all keyboard shortcut constants into separate files, make them tool separated and deduplicate them

MichaelBuessemeyer committed:
WIP fix imports and other compile errors

MichaelBuessemeyer committed:
improve & unify naming of tool shortcut constants

MichaelBuessemeyer committed:
try to migrate keyboardJS to Keystrokes

MichaelBuessemeyer committed:
move tool related shortcuts to their controller definitions & improve their typing
- Open TODO: plane_controller

MichaelBuessemeyer committed:
Migrate Plane controller; Annotation Tools is still WIP; fix TS

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
WIP. realize tool dependent keyboard shortcuts in ortho mode

MichaelBuessemeyer committed:
fix shortcuts using ctrl as keystroke requires it to be spelled control
- fix arrow keys as well
- fix default skeleton looped keyboard shortcuts
- fix generating tool dependant keyboard shortcuts

MichaelBuessemeyer committed:
remove debug statement

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
WIP reorganize keyboard shortcuts modal

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
WIP improve UI

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
refine key combo visualization column

MichaelBuessemeyer committed:
fix displayed ">" and format keys in recording modal properly

MichaelBuessemeyer committed:
fix imports

MichaelBuessemeyer committed:
fix imports and ts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
use event.key directly as stored variable to identify key as this is used by keystrokes as well. And do not store with intermediate names

MichaelBuessemeyer committed:
WIP: proper collision detection & var naming adjustments

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix some const definitions for shortcuts

MichaelBuessemeyer committed:
improve keyboard shortcut collision detection

MichaelBuessemeyer committed:
WIP add keyboard shortcut config persistence

MichaelBuessemeyer committed:
add missing renaming

MichaelBuessemeyer committed:
fix keyboard shortcut persistence

MichaelBuessemeyer committed:
sort imports

MichaelBuessemeyer committed:
do early collision detection in shortcut recorder modal

MichaelBuessemeyer committed:
wip diaplyed configured shortcuts in info tab

MichaelBuessemeyer committed:
WIP refine UI style of info tab icons

MichaelBuessemeyer committed:
improve key recording to register digits and lower case letters in case they are modified

MichaelBuessemeyer committed:
remove logging

MichaelBuessemeyer committed:
make keyboard shortcut order in table consistent

MichaelBuessemeyer committed:
add keyboard shortcuts modal to view dataset mode

MichaelBuessemeyer committed:
adjust info tab shortcuts rendering to state on master

MichaelBuessemeyer committed:
fix mini bug where meta key was shown in ui twice

MichaelBuessemeyer committed:
add mouse shortcuts to shortcuts modal

MichaelBuessemeyer committed:
move mouse shortcuts to separate file and deduplicate css and minor wording improvements

MichaelBuessemeyer committed:
use useWkSelector

MichaelBuessemeyer committed:
disable shortcut editing for not authenticated users

MichaelBuessemeyer committed:
fix imports

MichaelBuessemeyer committed:
add changelog

MichaelBuessemeyer committed:
merge master into this branch

MichaelBuessemeyer committed:
add check that no untyped useSelector is imported in the  repo. Use useWkSelector instead

MichaelBuessemeyer commented on tools/check-no-untyped-use-selector-imports.js:
this script is ai generated and seem sot work. Did not review this much as it is similarly designed to `./tools/check-no-untyped-select-imports.js`

MichaelBuessemeyer reviewed (commented):
So should be ready for the first round :D

MichaelBuessemeyer committed:
fix keyboard shortcut registering api test

MichaelBuessemeyer committed:
temp disable todo check

MichaelBuessemeyer commented on .github/workflows/check_todos.yml:
```diff
@@ -12,4 +12,4 @@ jobs:
       - uses: actions/checkout@v4
 
       - name: Check for TODO comments without issue links
-        run: tools/check-todos.sh
```
TODO undo

philippotto replied:
lol, why? couldn't you switch the pr state back to draft if you needed a dev build? maybe the docker image is even built independent of the todo-check (did not check this yet).

MichaelBuessemeyer replied:
I wanted the code rabbit review and it skips the review in draft state. Maybe I could have acchieved the same with a command but thats the way I chose now :see_no_evil:

MichaelBuessemeyer commented on frontend/javascripts/libs/input.ts:
```diff
@@ -79,138 +147,105 @@ const keyboard = new Keyboard(
       document.addEventListener("keydown", this.preventBrowserSearchbarShortcut);
-      this.attach(EXTENDED_COMMAND_KEYS, this.toggleExtendedMode);
-      // Add empty callback in extended mode to deactivate the extended mode via the same EXTENDED_COMMAND_KEYS.
-      this.attach(EXTENDED_COMMAND_KEYS, noop, noop, true);
-      for (const key of Object.keys(extendedCommands)) {
-        const callback = extendedCommands[key];
-        this.attach(key, callback, noop, true);
-      }
+      this.isPreventBrowserSearchbarShortcutActive = true;
     }
```
neat good find :+1:

MichaelBuessemeyer commented on frontend/javascripts/viewer/view/input_catcher.tsx:
```diff
@@ -148,6 +148,7 @@ function InputCatcher({
 
   const activeTool = useWkSelector((state) => state.uiInformation.activeTool);
 
+  // TODOM: how to handle this? This is not part of the "keyboard shortcut structure".
```
@philippotto Do you have an idea?
The code is:
```
// TODOM: how to handle this? This is not part of the "keyboard shortcut structure".
  const isShiftPressed = useKeyPress("Shift");
  const isControlOrMetaPressed = useKeyPress("ControlOrMeta");
  const isAltPressed = useKeyPress("Alt");

  const adaptedTool =
    viewportID === ArbitraryViews.arbitraryViewport
      ? AnnotationTool.SKELETON
      : viewportID === OrthoViews.TDView
        ? AnnotationTool.MOVE
        : adaptActiveToolToShortcuts(
            activeTool,
            isShiftPressed,
            isControlOrMetaPressed,
            isAltPressed,
          );
``` 
So will this just be constant or should this also be configurable? :thinking:

philippotto replied:
> So will this just be constant or should this also be configurable? 

no, this should not be configurable in my opinion. let's not go too crazy here :)

MichaelBuessemeyer committed:
WIP apply coderabbit feedback

MichaelBuessemeyer committed:
add test to check for unique keyboard handler ids

MichaelBuessemeyer committed:
make keyboard config schema parsing less aggressive

MichaelBuessemeyer committed:
fix key duplication and add missing key prop

MichaelBuessemeyer committed:
fix new mouse move icon

MichaelBuessemeyer committed:
hide classic controls shortcut hints when they are not active

MichaelBuessemeyer committed:
remove empty shortcut category

MichaelBuessemeyer committed:
do not auto fill shortcuts in case last binding is deleted

MichaelBuessemeyer commented on frontend/javascripts/libs/input.ts:
```diff
@@ -24,33 +30,97 @@ import { addEventListenerWithDelegation, isNoElementFocused } from "./utils";
+  onReleased?: KeyboardLoopFn;
+};
+// TODO: Improve type naming.
+export type KeyBindingMap = Record<KeyboardKeyOrCombo, KeyboardNoLoopHandler>;
```
The naming of types here is not perfect imo. Do you have improvement suggestions?

philippotto replied:
`KeyBindingToHandlerMap` is longer but more precise. I'm not totally sure here, either.

MichaelBuessemeyer replied:
Took `KeyComboToHandlerMap` as the key is now called `KeystrokesKeyCombo`

MichaelBuessemeyer commented on frontend/javascripts/libs/shortcut_component.ts:
```diff
@@ -3,27 +3,23 @@ import { useEffect } from "react";
+export default function Shortcut({ keys, onTrigger, supportLoop, supportInputElements }: Props) {
   useEffect(() => {
-    const keyboard = new (props.supportLoop ? InputKeyboard : InputKeyboardNoLoop)(
-      {
-        [props.keys]: props.onTrigger,
-      },
-      {
-        supportInputElements: props.supportInputElements,
-      },
-    );
+    const keyboard = supportLoop
+      ? new InputKeyboard({ [keys]: { onPressedWithRepeat: onTrigger } }, { supportInputElements })
+      : new InputKeyboardNoLoop({ [keys]: { onPressed: onTrigger } }, { supportInputElements });
 
     return () => {
       keyboard.destroy();
     };
-  }, [props.keys, props.onTrigger, props.supportLoop, props.supportInputElements]);
+  }, [keys, onTrigger, supportLoop, supportInputElements]);
 
   return null;
 }
```
This component could use the new keystokes hook. Should I implement this @philippotto?

philippotto replied:
I'm not sure what the new keystrokes hook is. can you link it? haven't seen it (yet).

MichaelBuessemeyer replied:
This here: https://github.com/RobertWHurst/Keystrokes/#using-keystrokes-with-react

Not sure how useful this is. Reads like a re-render on successful keyboard input to me

philippotto replied:
I think our wrapper is so lightweight that it's okay to keep it. We can swap it in the future, when we feel the need.

MichaelBuessemeyer committed:
resolve / comment TODO comments

MichaelBuessemeyer committed:
re-enable todo checking pipeline

MichaelBuessemeyer commented on frontend/javascripts/viewer/controller/viewmodes/plane_controller.tsx:
```diff
@@ -314,6 +156,8 @@ class PlaneController extends PureComponent<Props> {
   storePropertyUnsubscribers: Array<(...args: Array<any>) => any> = [];
   isStarted: boolean = false;
+  // TODOM: improve typing
+  unsubscribeKeyboardListener: any = () => {};
```
outdated already

MichaelBuessemeyer commented on frontend/javascripts/viewer/view/right_border_tabs/dataset_info_tab_view.tsx:
```diff
@@ -73,73 +80,111 @@ type State = {
+  keyboardShortcutsConfig: KeyboardShortcutsMap<string>,
+  isInPlaneMode: boolean,
+): ShortcutInfo[] => {
+  const toUiElement = (keyboardShortcutHandlerId: AnyKeyboardHandlerId) =>
+    (keyboardShortcutsConfig[keyboardShortcutHandlerId] ?? []).flatMap((keyCombo, comboIndex) => {
+      const capitalizedKeyCombo = keyCombo.map((keys) => keys.map((key) => key.toUpperCase()));
+      return keyComboChainToUiElements(
+        capitalizedKeyCombo,
+        true,
+        `${keyboardShortcutHandlerId}-${comboIndex}-`,
+      );
+    });
```
yeah thats how it was on the master. Having the alt key written in caps is correct. So should be fine in this specific case here :+1:. But not in general. So might need adjusting in the future or so

MichaelBuessemeyer reviewed (commented):
Ok, I think I applied all the feedback from coderabbit now.

Additionally your UI suggestions are also done and I fixed the icon I added.

So should be ready for review in case the CI is green / docker image is built

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

philippotto commented on frontend/javascripts/libs/input.ts:
```diff
@@ -24,33 +30,96 @@ import { addEventListenerWithDelegation, isNoElementFocused } from "./utils";
-type KeyboardKey = string;
-type MouseButton = string;
-type KeyboardHandler = (event: KeyboardEvent) => void | Promise<void>;
+export type KeyboardKeyOrCombo = string;
```
can you add examples for valid strings here?

MichaelBuessemeyer replied:
Good suggestion. I also renamed it to `KeystrokesKeyCombo` to make this clearer

philippotto commented on frontend/javascripts/libs/input.ts:
```diff
@@ -24,33 +30,96 @@ import { addEventListenerWithDelegation, isNoElementFocused } from "./utils";
-type KeyboardHandler = (event: KeyboardEvent) => void | Promise<void>;
+export type KeyboardKeyOrCombo = string;
+export type KeyboardHandlerFn = (event: KeyboardEvent) => void | Promise<void>;
 // Callable Object, see https://www.typescriptlang.org/docs/handbook/2/functions.html#call-signatures
```
the comment should be moved above the ` (arg0: number, isOriginalEvent:...` line.

philippotto commented on frontend/javascripts/libs/input.ts:
```diff
@@ -24,33 +30,96 @@ import { addEventListenerWithDelegation, isNoElementFocused } from "./utils";
+  finalKeyEvent: KeyEvent<KeyboardEvent, BrowserKeyComboEventProps>;
+};
+type KeystrokesHandler = (event: KeystrokesHandlerArgs) => void;
+type KeyboardBindingPress = {
```
now there's
- KeyBinding
- KeyboardBinding
- KeyStrokesHandler
- Keyboard*Handler
- KeyboardKeyOrCombo
- (maybe more?)

please try to unify these and add a docstring that explains the difference between the concepts that remain :)
it might make sense to put that into a design doc first so that we can have consensus before changing the code.

MichaelBuessemeyer replied:
Oh ups, sorry misread the last 2 sentences and already moved on and renamed the types into something that is more consistent and makes more sense. Moreover, I reordered them, grouped them by "these types are used to interface with the InputKeyboard classes" and "theses types are used by the InputKeyboard classes internally to interface with Keystrokes".

Hope this makes things much clearer now. If not, I'll open up a design doc 👍

philippotto commented on frontend/javascripts/libs/input.ts:
```diff
@@ -67,8 +136,6 @@ function shouldIgnore(event: KeyboardEvent, key: KeyboardKey) {
 // to the underlying KeyboadJS library to do its dirty work.
 // Pressing a button will only fire an event once.
-const EXTENDED_COMMAND_KEYS = isMac ? "command + k" : "ctrl + k";
-const EXTENDED_COMMAND_DURATION = 3000;
```
is there still this duration somewhere else?

MichaelBuessemeyer replied:
Nope, I think the duration / timeout between registered key sequences is managed by the Keystrokes library and the default settings worked well for me

philippotto commented on frontend/javascripts/libs/input.ts:
```diff
@@ -79,138 +146,105 @@ const keyboard = new Keyboard(
-
-    if (extendedCommands) {
-      this.hasExtendedBindings = true;
+    const hasLegacyExtendedKeyboardShortcut = Object.keys(initialBindings).some((keyCombo) =>
```
why legacy? can't the user create a new binding with ctrl + k which would require the same logic?

MichaelBuessemeyer replied:
well I thought legacy might be a good way to name this as ctrl + k is no longer required for a user to switch tools as this is now configurable. But it is still the default shortcut. 
But I renamed it now to `usesShortcutToFastFocusBrowserSearchbar` which makes more sense imo as it defines it "usage" / "purpose". What do you think? (Feel free to resolve)

philippotto commented on frontend/javascripts/test/misc/keyboard_shortcut_handler_ids.spec.ts:
```diff
@@ -0,0 +1,32 @@
+} from "viewer/view/keyboard_shortcuts/keyboard_shortcut_constants";
+import { describe, expect, it } from "vitest";
+
+describe("Keyboard shortcut handler IDs", () => {
```
nice tests 👍

MichaelBuessemeyer replied:
Handing over the nice to claude 🙈

Do you think I should also add a few tests checking whether the input.ts keyboard classes are working as expecte? This is pretty easily testable with Keystrokes as it has build in testing / mocking support

philippotto replied:
> Do you think I should also add a few tests checking whether the input.ts keyboard classes are working as expecte?

not sure how many trouble you had with the input class during development. if it was smooth sailing, I don't really see the need for additional specs right now, but in the end you can decide this better :)

philippotto commented on frontend/javascripts/viewer/controller/combinations/tool_controls.ts:
```diff
@@ -752,6 +990,30 @@ export class BoundingBoxToolController {
   }
 
+  static getNoLoopedKeyboardControls(): KeyboardShortcutNoLoopedHandlerMap<PlaneBoundingBoxToolNoLoopedKeyboardShortcuts> {
+    const handleReactToModifier = (event: KeyboardEvent) => {
```
handleReact sounds redundant?

MichaelBuessemeyer replied:
correct, and the old name wasn't very expressive / sounded confusing to just now. I renamed it to `handleBBoxMoveModifierPressed` which is much better imo

philippotto commented on frontend/javascripts/viewer/controller/combinations/tool_controls.ts:
```diff
@@ -1107,15 +1369,30 @@ export class AreaMeasurementToolController {
+        onPressed: () => {
+          const state = Store.getState();
+          const isProofreadingActive = state.uiInformation.activeTool === AnnotationTool.PROOFREAD;
+          if (isProofreadingActive) {
```
is this if still necessary?

MichaelBuessemeyer replied:
yes definitely. this is the only allowed colliding shortcut. See `acceptedCollisions` in `keyboard_shortcut_utils.tsx`. I added a comment here to cross reference this.

philippotto commented on frontend/javascripts/viewer/controller/combinations/tool_controls.ts:
```diff
@@ -1229,3 +1506,37 @@ const toolToToolController = {
+  [AnnotationTool.ERASE_BRUSH.id]: EraseToolController.getNoLoopedKeyboardControls(),
+  [AnnotationTool.FILL_CELL.id]: FillCellToolController.getNoLoopedKeyboardControls(),
+  [AnnotationTool.VOXEL_PIPETTE.id]: VoxelPipetteToolController.getNoLoopedKeyboardControls(),
+  [AnnotationTool.LINE_MEASUREMENT.id]: LineMeasurementToolController.getNoLoopedKeyboardControls(),
```
couldn't this code (until the rest of the file) be removed with two loops or maps? `toolToToolController` should help here. typing might need some additional annotations or casts, but I'd prefer this over this long verbose block.

philippotto commented on unreleased_changes/9081.md:
```diff
@@ -0,0 +1,5 @@
+### Added
+- Added a modal displaying current keybaord and mouse shortcuts to WEBKNOSSOS. The keyboard shortcuts are editable.
```
I'd lead with the banger feature ;)

```suggestion
- Keyboard shortcuts are configurable now. They are displayed in a new shortcuts dialog (together with mouse shortcuts which are not editable).
```

philippotto commented on frontend/stylesheets/main.less:
```diff
@@ -499,4 +493,53 @@ body {
\ No newline at end of file
+}
+
+.shortcut-table-modal {
```
wouldn't it make more sense in the tracing less?

MichaelBuessemeyer replied:
hmm yeah maybe. My inital thought was that the modal is also available in view mode, but the less is active regardless :+1:

philippotto commented on frontend/javascripts/viewer/merger_mode.ts:
```diff
@@ -532,8 +532,10 @@ export async function enableMergerMode(
   unregisterKeyHandlers.push(
-    api.utils.registerKeyHandler("9", () => {
-      changeOpacity(mergerModeState);
+    api.utils.registerKeyHandler("9", {
```
what happens when the user configured 9 to something else? both are executed then? I'm wondering whether it would be more elegant to have a "specifity approach" (similar to CSS) so that the binding with the highest specifity wins. here in that case, the merge modal shortcut would have a very high specifity. 

alternatively (for this concrete shorcut), could it be handled like all the other mode-specific shortcuts?

MichaelBuessemeyer replied:
Oh didn't find this. Good find 👍

> What happens when the user configured 9 to something else? both are executed then? 

Both should be executed and the collision wouldn't be detected in the shortcuts modal. 

> I'm wondering whether it would be more elegant to have a "specifity approach" (similar to CSS) so that the binding with the highest specifity wins. here in that case, the merge modal shortcut would have a very high specifity.

I'd say this would be a new concept that would need more implementation effort. Moreover, we need to define the priority for each shortcut. Additionally, this would no longer make it possible to map two functions into one shortcut which might be useful power feature for users 🤔?

I could invest some time to map this into the "normal tools shortcuts". But do you think this is worth it in this iteration? I'd say this isn't much of a problem as users using the merge mode should be aware of the 9 shortcut, or not? Might not be worth the time for now imo.

philippotto commented on frontend/javascripts/viewer/controller.tsx:
```diff
@@ -52,13 +61,23 @@ type State = {
     gotUnhandledError: false,
     organizationToSwitchTo: null,
   };
+  unsubscribeKeyboardListener: any = () => {};
```
can we do better than any?

MichaelBuessemeyer replied:
`() => void` would be the correct type. But I used `ReturnType<typeof listenToStoreProperty>` as it is more reliable in case the return type of `listenToStoreProperty` changes. What do you think?

philippotto commented on frontend/javascripts/viewer/controller.tsx:
```diff
@@ -208,26 +228,85 @@ class Controller extends PureComponent<PropsWithRouter, State> {
+      };
+
+    // Wrapped in a function to ensure each time the map is used, a new instance of getHandleToggleSegmentation
+    // is created and thus "leastRecentlyUsedSegmentationLayer" not being shared between key binding maps.
```
I think the comment is at the wrong place?

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/shortcut_recorder_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+  );
+};
+
+const SampleKeyCombo: KeyboardComboChain = [["Control", "a"], ["o"]];
```
capslock case for constant. also, the name is keycombo but the type keyboardcombochain which is not consistent.

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/shortcut_recorder_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+    if (keyComboChain.length > 0) {
+      onSave(keyComboChain);
+    } else {
+      // nothing recorded -> treat as cancel (or you can choose to save empty)
```
either cancel or save. the comment is misleading

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/shortcut_recorder_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+        <Text type="secondary">
+          Press keys now. Release all keys to finish the current stroke. Press keys again to add a
+          subsequent stroke. Example: {keyComboChainToUiElements(SampleKeyCombo, false)}. Reset
+          everything with Esc + Ctrl.
```
modifiers usually come first.

```suggestion
          everything with Ctrl + Esc.
```

also, these should probably be formatted somehow so that + is not interpreted as a key.

MichaelBuessemeyer replied:
current version:
<img width="940" height="164" alt="image" src="https://github.com/user-attachments/assets/970826b9-f375-4ef0-8f2b-6b8c78177861" />

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/shortcut_recorder_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+      <div style={{ padding: 8 }}>
+        <CollisionWarningAlert shortcutCollisions={shortcutCollisions} />
+        <Text type="secondary">
+          Press keys now. Release all keys to finish the current stroke. Press keys again to add a
```
I'd not focus the sequence stuff too much. it's quite niche-y in my opinion and should stay that way.

sth like:
```
Press a keyboard combination now to record it. You can also record a sequence of keys. Example: ...
```

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/shortcut_recorder_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+              <div style={{ width: "100%", overflow: "auto" }}>
+                <Text strong>Recorded:</Text>{" "}
+                <span style={{ marginLeft: 8 }}>
+                  {keyComboChainToUiElements(keyComboChain, false) || "— waiting —"}
```
```suggestion
                  {keyComboChainToUiElements(keyComboChain, false) || "— waiting for user input —"}
```

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+    case "Meta":
+      return (
+        <>
+          <MacCommandOutlined />/<WindowsOutlined />
```
should we use `isMac()` to make this dynamic?

MichaelBuessemeyer replied:
awesome, very nice suggestion 🎉

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  }
+}
+
+function escapeReservedKeystrokeCharacters(key: string): string {
```
what happens with the `\\`? why is it necessary?

MichaelBuessemeyer replied:
Well it escapes the revered keystore characters. `+` becomes `\+`. The "\\" is needed as "\" is already a reservered js char in strings. "\\" maps to the actual string "\"

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+
+  for (const m of modifiersOrder) {
+    if (seen.has(m)) presentModifiers.push(m);
+  }
```
(untested)

```suggestion
  const presentModifierSet = new Set(combo).intersection(new Set(MODIFIER_KEYS));
  const nonModifiers = Array.from(new Set(combo).difference(new Set(MODIFIER_KEYS)));
  const presentModifiers = modifiersOrder.filter(m => presentModifierSet.has(m));
```

MichaelBuessemeyer replied:
I tested this and it did not show differences, but I'd say 
`const nonModifiers = Array.from(new Set(combo).difference(new Set(MODIFIER_KEYS))); ` 
does not guarantee to keep the original order.

So I went for:
```
const deduplicatedCombo = uniq(combo);
const presentModifierSet = new Set(deduplicatedCombo).intersection(MODIFIER_KEYS);
const nonModifiers = deduplicatedCombo.filter((key) => !MODIFIER_KEYS.has(key));
const presentModifiers = modifiersOrder.filter((m) => presentModifierSet.has(m));
```

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+    .join(" + ");
+}
+
+export function keyComboChainToKeystrokesConfig(comboChain: KeyboardComboChain): string {
```
"config string" might be better?

MichaelBuessemeyer replied:
that would be inconsistent with the other namings, so I'd vote against this or we start renaming the other types and variables as well 🤔

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+
+export function keyComboChainToUiElements(
+  comboChain: KeyboardComboChain,
+  useFancy: boolean,
```
please add a comment what "fancy" means in this context.

MichaelBuessemeyer replied:
I renamed the parameter and added a comment 👍

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+    }),
+  ).filter((mapping) => mapping != null);
+  return Object.fromEntries(mappedShortcuts);
+};
```
the implementation is identical to the function above (except for param and return type). can you unify this? e.g., with a generic.

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  return Object.fromEntries(mappedShortcuts);
+};
+
+function keyComboChainToSetArray(comboChain: KeyboardComboChain): ComparableKeyComboChain {
```
why "to set array" when you already have the term `ComparableKeyComboChain`?

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+export function keyboardShortcutMapToCollidingTuples<T extends string>(
+  config: KeyboardShortcutsMap<T>,
+): [ComparableKeyComboChain, T[]][] {
+  const result: [ComparableKeyComboChain, T[]][] = [];
```
please explain how to interpret the return type. I'm confused because result is being pushed to even when existingEntry is null? also, an object might be better than a tuple, since it would have readable property keys?

MichaelBuessemeyer replied:
the goal of this function is to give a map in form of `Record<KeySequence, KeyboardShortcutId[]>`. Meaning for each given key sequence in the config I know what shortcutIds map to this sequence. Meaning if there are more then 1 entry per sequence, there is a collision.

The problem is that a `KeySequence` is not a valid key for an object. But for the caller side, an object would make more sense imo.
=> I refactored the function to return a Map instead. However, the map is only returned. Internally, the tuple list still needs to be used in order to be properly find equal key sequences. Hopefully the comments explain this sufficiently

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  return result;
+}
+
+function buildToolDependentNoLoppedHandler(
```
lopped -> looped

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  };
+}
+
+function buildToolDependentLoppedHandler(
```
same here

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+    onReleased: (...args: Parameters<KeyboardLoopFn>) => {
+      const activeToolId = Store.getState().uiInformation.activeTool.id;
+      toolToHandlerMap[activeToolId]?.onReleased?.(...args);
+    },
```
I feel like this whole looped/non-looped distinction adds lots of verbose code. couldn't be both be in the same toolToHandlerMap and the code here would simply check whether `onPressedWithRepeat` and `onPressed` exist each?

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+        toolOnlyShortcuts[handlerId] = config[handlerId];
+      }
+    }
+  }
```
some comments would help. the rest of the function is identical to its sibling function (except for some types). another point for eliminating loop vs no-loop.

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  return bindings;
+};
+
+type ComparableKeyComboChain = Set<string>[];
```
I would prefer "Sequence" over "Chain" but maybe that's just a matter of taste.

MichaelBuessemeyer replied:
yeah good suggestion 👍

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  conflictingHandlerIds: string[];
+};
+
+function buildParentMap(hierarchy: Record<string, string[]>): Record<string, string | null> {
```
docstrings would help

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -0,0 +1,493 @@
+  return parentMap;
+}
+
+function getCollidableEntities(
```
did not review the rest of the file since it's a lot of code without any high-level explanations.

philippotto commented on frontend/javascripts/viewer/controller/viewmodes/plane_controller.tsx:
```diff
@@ -717,11 +535,11 @@ class PlaneController extends PureComponent<Props> {
-    this.unsubscribeStoreListeners();
   }
 
   createToolDependentKeyboardHandler(
```
unused now?

MichaelBuessemeyer replied:
ah nice 👍 good find 🎉

the collision handling is now done by `buildKeyBindingsFromConfigAndLoopedMappingForTools`.

philippotto reviewed (changes requested):
ok, I'm giving up for now 🙈 I didn't make it through all files. it's a lot of code and I'm having a hard time internalizing all concepts and mechanisms. feel free to have a look at my comments, but no need to act directly, as I'd like to take a step back.

I might be in the wrong, but I hope that we can simplify the existing architecture quite a bit. To name a few suggestions I have:

- loop, no-loop and delay should be unified so that we don't need similar code three-times. we can still have two keyboard instances for loop/noloop, but the declarative definition of the shortcuts should not get convoluted by it. the action binding already differentiates via `onPressedWithRepeat` and `onPressed`. theoretically, delay could be encoded similarily in the controllers or at the shortcut definition (see below).
- the (collision) domain/entity stuff feels quite complicated. currently, there is KeyboardShortcutDomain, KeyboardShortcutCollisionEntityName and KeyboardShortcutCollisionHierarchy which does not feel intuitive. I'd do this with a "nested" domain string (see below).

In my head, my envisioned definition of the shortcuts would be sth like this:

```ts
// The id is unique across all shortcuts. The user config attaches bindings to the shortcut. Controller (or other components)
// attach actions to the shortcut.
type ShortcutID = "SWITCH_VIEWMODE_PLANE" | ...;

// Domains can be nested using the dot notation. From that a hierarchy follows which allows to check for conflicts. Another dictionary for human readable names is probably necessary.
type ShortcutDomain = "GENERAL" | "GENERAL.PLANE_MODE" | "GENERAL.PLANE_MODE.SKELETON";

// ShortcutDefinitions provide multiple properties by which can be filtered to get the desired subsets.
class ShortcutDefinition {
  constructor(
    public ID: ShortcutID,
    public description: string,
    public defaultBindings: KeyboardComboChain[],
    public domain: ShortcutDomain,
    public options?: { delay: boolean },
  ) {}
}

// The following is a flat list of all shortcuts in WK.
export const configurableShortcuts = [
  new ShortcutDefinition("SWITCH_VIEWMODE_PLANE", "View in plane mode", [["Shift", "1"]], "GENERAL"),
  new ShortcutDefinition("MOVE_MULTIPLE_FORWARD", "Fast move forward", [["Shift", "f"]], "GENERAL.PLANE_MODE", {
    delay: true,
  })
  // ...
];
```

Some details might be missing (e.g., I might have tool-specific stuff seem too easy above), but I feel like this is a good basis. my suggestion probably has less extensive typing in contrast to your code, but I feel like that is okay. by the way, I'm not a huge fan of enums. 

what do you think? I think, a good next step would be to create a notion doc that explains the architecture, concepts and terminology (important point: what is a shortcut, binding, ...) of this PR. please either try to use my suggestion above as a basis or if you think that it doesn't make sense, write down the current (or an improved) version of your envisioned architecture. using that doc, we can discuss further :)

MichaelBuessemeyer committed:
Merge branch 'master', commit 'e13b6240a7fd1862a938194dd14217f0749b592a' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer commented on frontend/javascripts/libs/input.ts:
```diff
@@ -238,82 +272,93 @@ export class InputKeyboard {
+        delete this.keyCallbackMap[keyCombo];
+      }
 
-        if (delayTimeoutId != null) {
-          clearTimeout(delayTimeoutId);
-          delayTimeoutId = null;
-        }
-      },
-      false, // preventRepeatByDefault
-    ];
-    keyboard.withContext("default", () => {
-      keyboard.bind(...binding);
-    });
+      if (delayTimeoutId != null) {
+        clearTimeout(delayTimeoutId);
+        delayTimeoutId = null;
+      }
+      const event = findEventInKeystrokeComboEvent(keyEvents, finalKeyEvent);
+      if (onReleased != null && event != null) {
+        onReleased(1, true, event);
+      }
+    };
```
@philippotto do you think this is a relevant problem? Should I guard against this like suggested here?

MichaelBuessemeyer committed:
apply some feedback

MichaelBuessemeyer committed:
unify looped and no looped keyboard

Claude implementation plan overview:

 Context

 input.ts exports two keyboard classes with fundamentally different handler signatures.
 Every layer above them — type aliases, builder functions, per-controller handler methods —
 duplicates itself to track the loop/no-loop distinction. The goal is to collapse that split
 all the way up the stack so that:

 - Only one InputKeyboard class exists.
 - Only one builder function per shape (regular vs. per-tool) exists.
 - Each controller has one handler-map method and one InputKeyboard instance.
 - The looped? field in KeyboardShortcutMetaInfo (currently unused at runtime) disappears.
 - Whether a shortcut loops is determined solely by handler shape: onPressed = no-loop,
 onPressedWithRepeat = loop.
 - The keyboard delay is maintained automatically — the class subscribes to
 userConfiguration.keyboardDelay in the store and keeps its internal delay in sync.
 Call sites only need delayed: true on individual handlers; no delay wiring needed.

MichaelBuessemeyer commented on frontend/javascripts/viewer/controller/viewmodes/plane_controller.tsx:
```diff
@@ -438,7 +277,121 @@ class PlaneController extends PureComponent<Props> {
+    const controllerBindings = buildKeyBindingsFromConfig(
+      keyboardShortcutsConfig,
+      this.getHandlerMap(),
+    );
+    const toolBindings = buildKeyBindingsFromConfigForTools(
+      keyboardShortcutsConfig,
+      AllToolKeyboardControls,
+    );
+
+    this.input.keyboard = new InputKeyboard({
+      ...controllerBindings,
+      ...toolBindings,
+      // Enter & Escape are constant and not configurable.
+      enter: {
+        onPressedWithRepeat: (_, _isOriginalEvent, event) => Store.dispatch(enterAction(event)),
+      },
+      esc: { onPressedWithRepeat: () => Store.dispatch(escapeAction()) },
+    });
```
@philippotto on master / previously the enter / esc actions were looped. This is also the current behaviour. But I think the comment here makes sense. Why should this action be looped. I cannot think of a proper use case for this :thinking:

philippotto replied:
yeah, I agree.

MichaelBuessemeyer committed:
polish InputKeyboard refactoring and some type renaming

MichaelBuessemeyer committed:
implement discussed refactoring in accordance to design doc
- doc: https://www.notion.so/scalableminds/Design-Doc-Configurable-Keyboard-Shortcuts-Architecture-34bb51644c63806e9717f96af0aa62b9?d=34bb51644c63802cbacc001c20b65f56#34bb51644c63807eb167ea6563284892

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix ui text and space key recording

MichaelBuessemeyer committed:
fix some ui & proper recorder modal collision detection

- The recorder collision detection did not work hierarchy wise downwards,
now it does by also traverse the hierarchy down the collision tree when collecting all collision domains

MichaelBuessemeyer committed:
make comment tab refresh keyboard shortcuts on shortcut change

MichaelBuessemeyer committed:
properly reload edited keyboard shortcut in recorder modal

MichaelBuessemeyer committed:
move css and update changelog entry

MichaelBuessemeyer committed:
move theme provider to resolve cyclic imports

MichaelBuessemeyer committed:
apply various review feedback

MichaelBuessemeyer committed:
do not initially mark edited shortcut as a collision & do not allow to add same shortcut multiple times to one action

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix detecting spacebar as input

MichaelBuessemeyer reviewed (commented):
Ok after a lot of refactoring  & feedback applying & bug fixing I can finally present the second iteration of this PR 🎉 

Happy to hear your newest feedback 👂

MichaelBuessemeyer commented:
@coderabbitai full-review

MichaelBuessemeyer commented:
@philippotto, please wait until I have applied the newest feedback. Seems like quite some good things were found

MichaelBuessemeyer committed:
clean up scrolling via space prevention

MichaelBuessemeyer committed:
clean up keystrokes test instance after test

MichaelBuessemeyer committed:
make return type of shortcut fetching route more accurate

MichaelBuessemeyer committed:
fix general editing shortcuts

MichaelBuessemeyer committed:
fix action wording and remove trailing spaces

MichaelBuessemeyer committed:
make  tool dependant handler pass correct arguments to on release handler when not a looped handler

MichaelBuessemeyer committed:
hot reload layouting shortcuts on change

MichaelBuessemeyer committed:
save guard recorder modal against rapid updates & fix css

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix formatting & organize imports

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_config_modal.tsx:
```diff
@@ -0,0 +1,499 @@
+    setLocalShortcutConfig(getAllDefaultKeyboardShortcuts());
+  };
+
+  const shortcutsTabItems: TabsProps["items"] = [
```
this structure feels like it's repeating the structure that is already defined elsewhere. would it be easy to dry that?

philippotto replied:
did this get lost? I don't think it's critical, but if it's easy...

MichaelBuessemeyer replied:
I don't see how to simply this to be honest. The order in what table is in what tab does not completely correlate to anything already present. Moreover, the MouseShortcut Tables are included here as well.

Or do you see a way to simply this?

philippotto replied:
alrighty then, I trust your judgement :ok_hand:

philippotto reviewed (changes requested):
awesome feature! I like the new structure way better 👍 code-wise I don't have any feedback anymore, but I have to admit that I didn't read every line due to the PR size. I mainly looked at the architecture and spot checked stuff.

now, for the testing feedback (feel free to disagree):
- [x] the default shortcuts to switch view modes did not work for me (Shift + 1/2/3)
- [x] "View in plane arbitrary mode" and "View in plane arbitrary plane mode" have the word "plane" once too often?
- [ ] toggling skeleton trees only works in the skeleton tool now. okay for me, but should be noted in the changelog probably.
- [x] if easy, I'd swap the columns for name <> binding so that name comes first.
- [x] Saving an unedited shortcut removes it from the list :thinking:
- [x] Rename everything arbitrary related to "flight"!
- [x] inputting the existing binding for the current shortcut, still shows it as a conflict (the conflict msg doesn't mention the shortcut name though)
- [x] in modal title: "Keyboard Shortcut Configuration" but the mouse Shortcuts are listed too. should be "Shortcut Configuration" probably?
- [x] plus-icon should have tooltip ("add additional keyboard binding"). same for edit and X button.
- [x] when I delete a keybinding, it looks a bit weird (there's only the restore button). How about "No shortcut configured" in italics + restore-button (without the "restore default" text; this can be moved into a tooltip)
- [x] mention that mouse shortcuts are not editable
- [x] Arbitrary mode is the second tab but it's rarely used. I'd move this to the back of the tabs. also, rename it to "flight", as we dont use arbitrary in the UI afaik
- [x] would it be simple to have tooltips for the different mouse-icons? I think it might not be obvious for everyone what they mean exactly.
- [x] "Mesh Click Related Shortcuts Mouse Shortcuts" <- doppelt gemoppelt
- [x] "Navigation in Arbitrary Mode Shortcuts" <- sounds a bit weird. maybe simply omit shortcuts? the entire modal is about shortcuts. I think, it's fair to remove that from all the subheadings. for mouse shortcuts, the headings could use "... (mouse-specific)"
- [x] I'd rename the tab title "Plane Mode Tools" to "Plane Mode - Tool Specific Functionality". alternatively, merge the two tabs (the first doesn't have that many entries, anyway).

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
remove oblique mode from newly created annotations and tasks

MichaelBuessemeyer committed:
switch columns in shortcut table

MichaelBuessemeyer committed:
remove shortcut allowing to switch to oblique mode

MichaelBuessemeyer committed:
improve shortcut modal title

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
rename arbitrary to flight mode and move on shortcut recorder modal event handler functions

MichaelBuessemeyer committed:
fix detecting self overlapping collisions

MichaelBuessemeyer committed:
add tooltips to shortcut modal buttons

MichaelBuessemeyer committed:
format

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
refine restore keyboard shortcut

MichaelBuessemeyer committed:
add info to mouse shortcuts that they are not editable via info icon

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
move flight mode tab in shortcut modal

MichaelBuessemeyer committed:
add explanations to mouse icons via tooltip

MichaelBuessemeyer commented:
>  toggling skeleton trees only works in the skeleton tool now. okay for me, but should be noted in the changelog probably.

Ok I think this should be moved to general plane shortcuts again :+1:. Was my mistake. Sorry. Or do you think it makes more sense in skeleton tool only? Because toggling the segmentation is globally possible as well :thinking:. So toggling tree visibility should be as well, or not?


> I think, it's fair to remove that from all the subheadings. for mouse shortcuts, the headings could use "... (mouse-specific)"

Ok, done.

> I'd rename the tab title "Plane Mode Tools" to "Plane Mode - Tool Specific Functionality". alternatively, merge the two tabs (the first doesn't have that many entries, anyway).

Ok, I merged them and gave the parts sub headings. This is now unique to the tab. So not sure how perfect of a solution this is :thinking:

MichaelBuessemeyer committed:
merge two plane mode tabs and name mouse shortcuts "(mouse-specific)" in UI

MichaelBuessemeyer committed:
Try to keep shortcuts to base keys of the shortcut and not to the modified keys and also display them correctly in the ui

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix schema versioning after merge

MichaelBuessemeyer committed:
store unmodified keyboard  layout map in local storage and move it to redux store

MichaelBuessemeyer committed:
fix mouse shortcuts visually

MichaelBuessemeyer committed:
fix FE tests

MichaelBuessemeyer committed:
remove oblique from allowed modes in annotation snapshots

MichaelBuessemeyer committed:
sort imports

MichaelBuessemeyer committed:
fix storing layout map

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix typo in changelog from other pr

MichaelBuessemeyer committed:
update docs

MichaelBuessemeyer committed:
remove react-keystrokes

MichaelBuessemeyer commented:
So, I applied all the feedback. Except for 
>> toggling skeleton trees only works in the skeleton tool now. okay for me, but should be noted in the changelog probably.

> Ok I think this should be moved to general plane shortcuts again 👍. Was my mistake. Sorry. Or do you think it makes more sense in skeleton tool only? Because toggling the segmentation is globally possible as well 🤔. So toggling tree visibility should be as well, or not?

What do you think about making the skeleton visibility shortcut available everywhere again?

Moreover, the solution discussed here is implemented: https://scm.slack.com/archives/C5AKLAV0B/p1777907953747809

Could you please give this another review? I re-deployed the dev instance for testing

philippotto commented:
> What do you think about making the skeleton visibility shortcut available everywhere again?

Yes, I think, this makes sense, because it's been like that before.

philippotto commented on docs/ui/keyboard_shortcuts.md:
```diff
@@ -4,6 +4,32 @@ The most important shortcuts are always shown in the status bar at the bottom of
+
+    **Chrome and Edge** expose this information via the [Keyboard Layout API](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard_API){target=_blank} and always show the correct character label immediately.
+
+    **Firefox, Safari, and other browsers** do not support this API. WEBKNOSSOS works around this limitation by learning your layout at runtime: The first time you press a symbol key **without any modifier** (no ++shift++, ++ctrl++, etc.), the correct label is detected and stored for your session and future sessions in your browser. Until a key has been pressed at least once this way, it falls ba ...
```
good that you documented this here (also for us devs) 👍

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_layout_utils.ts:
```diff
@@ -0,0 +1,156 @@
+
+const LOCAL_STORAGE_KEY = "keyboardLayoutMap";
+
+// US layout fallback for common sign/punctuation key codes.
```
I havent read everything yet, but it would be nice to have a high-level explainer (like in the docs?) in the module here first. if that already exists somewhere else, you can link to it here. linking to the docs would also be ok probably.

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_layout_utils.ts:
```diff
@@ -0,0 +1,156 @@
+// Used when navigator.keyboard.getLayoutMap() is unavailable and the key has
+// not yet been seen unmodified in the current session.
+const US_LAYOUT_FALLBACK: Record<string, string> = {
+  Minus: "-",
```
what event property is used to index into this dict?

MichaelBuessemeyer replied:
`KeyboardEvent.code`. I added this to the comment

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_layout_utils.ts:
```diff
@@ -0,0 +1,156 @@
+  if (isKeyboardLayoutApiAvailable()) {
+    const successful = await loadFromLayoutAPI();
+    if (successful) {
+      // Still listen to keyboard input to detect layout changes.
+      document.addEventListener("keydown", registerKeyForLayoutMap, { capture: true });
```
is this cheap? (I'm thinking of rapidly pressing `f`. if I keep pressing a key, it won't trigger repeatedly, right?)
if not, maybe it's ok to tell users to reload the page if they change their layout. should happen very rarely anyway. alternatively, maybe at least introduce some kind of throttling (only re-check layout every minute?).

MichaelBuessemeyer replied:
well not sure what you mean, but the `registerKeyForLayoutMap` is executed on every keydown event. I now added a guard to prevent considering repeated events when the user keeps a key pressed.
> if not, maybe it's ok to tell users to reload the page if they change their layout. should happen very rarely anyway. 

And how should we communicate this? If we do not have a mechanism to tell them?


Moreover, the overhead of the permanent listening is there anyway in case of firefox or so

philippotto replied:
ok, it's probably fine :crossed_fingers:

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_layout_utils.ts:
```diff
@@ -0,0 +1,156 @@
+}
+
+// Called from the shortcut recorder modal's keydown handler (all browsers) and the global
+// keydown listener registered above (non-API browsers only).
```
non-api browsers only isn't correct because of "// Still listen to keyboard input to detect layout changes." I think

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_types.ts:
```diff
@@ -48,6 +48,13 @@ export type KeyCombination = string[];
 export type KeyboardShortcutsMap = Record<KeyboardShortcutId, KeySequenceAlternatives>;
+// Stores a mapping from US-Keyboard key to key on the currently active layout.
+// The mapping is only used for visual display as it is not guaranteed to be complete.
+export type UnmodifiedLayoutMap = Map<string, string>;
```
please explain the term "unmodified" in this context. also an example would be nice.

MichaelBuessemeyer replied:
Maybe "KeyboardEventCodeToUnmodifiedKeyMap" is a more accurate name? I know it is long, but it is what the map does. I also adjusted the comment here to give more explanation and references the longer doc string in `keyboard_layout_utils.ts`

philippotto replied:
I'm fine with the name, as it's shorter. Just wanted to have an explanation for the term "unmodified" :) it's your call, if you still want to rename to KeyboardEventCodeToUnmodifiedKeyMap (however, I'm rather leaning against it due to the lenghtiness).

philippotto commented on frontend/javascripts/viewer/view/keyboard_shortcuts/keyboard_shortcut_utils.tsx:
```diff
@@ -25,13 +26,33 @@ import {
+    <Text type="warning" style={{ fontSize: 12 }}>
+      Your browser does not support the Keyboard Layout API (Chrome and Edge do). Labels for
+      punctuation and symbol keys fall back to US layout until you press those keys without a
+      modifier — after that, the correct label for your layout will be shown.
```
nit: if the user already uses the us layout, the labels will be correct before that.

philippotto commented:
didn't test everything (probably too much anyway) but noticed this:

- I'd rename "Shortcut Collisions Detected" to "Found Conflicting Shortcuts".
- Ctrl+K, S (etc) for switching to another tool don't work for me

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
make toggling trees shortcut globally available (as on master)

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
add first version of keyboard_layout_utils.ts docstring

MichaelBuessemeyer committed:
improve doc string

MichaelBuessemeyer committed:
apply review feedback

MichaelBuessemeyer committed:
fix typing in FE

MichaelBuessemeyer committed:
add missing tool switching shortcuts

MichaelBuessemeyer committed:
switch to keyboard shortcuts fork

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer reviewed (commented):
Thanks for you review.  The fork of Keystrokes is implemented and working See discussion https://scm.slack.com/archives/C5AKLAV0B/p1779286134443279 on this.

Moreover, I tested alls shortcuts and they are working implemented your feedback. Please have another look at this PR

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

Additionally, remove forward only method in UserService.scala and move messages of this PR into new Msg object

# Conflicts:
#	app/models/user/UserService.scala
#	conf/messages

MichaelBuessemeyer reviewed (commented):
CI seems to be broken currently. Will retry tomorrow or so

MichaelBuessemeyer committed:
Merge branch 'master' into custom-keyboard-shortcuts

philippotto commented on frontend/javascripts/test/snapshots/debug_htmls/.gitkeep:
```diff
@@ -1,3 +0,0 @@
-This folder can be used to dump snapshots during testing.
```
why deleted?

MichaelBuessemeyer replied:
was a mistake

philippotto commented on frontend/javascripts/test/snapshots/type_check/.gitkeep:
```diff
@@ -1,3 +0,0 @@
-This folder is used to write backend objects that are retrieved during end-to-end testing to disk.
```
why deleted?

philippotto reviewed (commented):
testing went well 👍 only noticed that there is no shortcut for activating the fill-tool and the measurement tools? the first is mentioned in the docs (so, it's a regression). the others are not mentioned but would still make sense? at least make them configurable (if no default shortcut makes sense, the binding could be empty right?)

MichaelBuessemeyer committed:
restore gitkeep files

MichaelBuessemeyer committed:
rename KeyboardEventCodeToUnmodifiedKeyMap back to UnmodifiedLayoutMap

MichaelBuessemeyer committed:
add missing tool switching shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer reviewed (commented):
Oh good catch. I added the missing tool switching shortcuts :+1:

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

# Conflicts:
#	schema/schema.sql

MichaelBuessemeyer committed:
Fix schema versioning

philippotto reviewed (approved):
🚀

MichaelBuessemeyer commented:
@fm3 could you please review the backend? Shouldn't take much time I think

fm3 commented on schema/schema.sql:
```diff
@@ -505,6 +505,14 @@ CREATE TABLE webknossos.user_datasetLayerConfigurations(
 
 
+CREATE TABLE webknossos.user_keyboardShortcutsConfigs(
+  _user TEXT CONSTRAINT _user_objectId CHECK (_user ~ '^[0-9a-f]{24}$') NOT NULL,
```
should this be per multiuser maybe?

fm3 commented on schema/evolutions/reversions/162-keyboard-shortcuts.sql:
```diff
@@ -0,0 +1,9 @@
+
+UPDATE webknossos.releaseInformation SET schemaVersion = 161;
+
+COMMIT TRANSACTION;
```
after merging the master into here you’ll need to bump the schema version, as we’re already on 164 now. You can try using `./tools/bump-evolution.sh`

MichaelBuessemeyer replied:
I  think the `./tools/bump-evolution.sh` has a bug. It now happend the 2nd time. On the import url fix pr (https://github.com/scalableminds/webknossos/pull/9516) I used this script and on this PR as well. In both cases the changes to schema.sql in this pr were removed and I had to restore them manually :/

fm3 replied:
Huh, that’s strange! If you can reproduce that bug, maybe you could have a quick look at fixing it? Otherwise I’ll try it soon.

MichaelBuessemeyer replied:
Hmm ok sorry for the false alarm. I tried again and this time everything worked perfectly. I only did not use the auto merge feature of IntelliJ. At least I think I used it each time. Not sure whether that's the actual cause, but as it works now, I'd say there is nothing to do here :+1:
Maybe better proofread the auto merge result from IntelliJ :thinking:

fm3 commented on util/src/main/scala/com/scalableminds/util/Msg.scala:
```diff
@@ -695,6 +695,8 @@ object Msg {
       val invalid: String = "Could not parse configuration."
       val invalidForDataset: String = "Could not parse dataset configuration."
+      val invalidKeyboardShortcutsConfig = "Keyboard Shortcuts Configuration is invalid"
+      val updatedKeyboardShortcutsConfig = "Keyboard Shortcuts Configuration updated"
```
```suggestion
      val updatedKeyboardShortcutsConfig = "Keyboard shortcuts configuration was updated."
```

fm3 commented on util/src/main/scala/com/scalableminds/util/Msg.scala:
```diff
@@ -695,6 +695,8 @@ object Msg {
       val updateSuccessForDataset: String = "Dataset configuration was successfully updated."
       val invalid: String = "Could not parse configuration."
       val invalidForDataset: String = "Could not parse dataset configuration."
+      val invalidKeyboardShortcutsConfig = "Keyboard Shortcuts Configuration is invalid"
```
```suggestion
      val invalidKeyboardShortcutsConfig = "Keyboard shortcuts configuration is invalid."
```

fm3 commented on app/controllers/ConfigurationController.scala:
```diff
@@ -35,6 +37,22 @@ class ConfigurationController @Inject()(
+    } yield addNoCacheHeaderFallback(Ok(shortcuts))
+  }
+
+  def updateKeyboardShortcutsConfig(): Action[JsValue] = sil.SecuredAction.async(parse.json(maxLength = 204800)) {
```
should be an Action[JsObject] and use json.validate[JsObject] directly (even though the other functions in this file still use the old style)

MichaelBuessemeyer replied:
But I simply reply with an message / string an not a full object. With `Action[JsObject]` it does not compile because `JsonOk(Msg.User.Configuration.updatedKeyboardShortcutsConfig)` is not a JsObject. Or should I convert it to one?

MichaelBuessemeyer replied:
Other functions use `Action[JsValue]` as well in this file

fm3 commented on conf/messages:
this file was deleted on master, is it re-created here?

MichaelBuessemeyer replied:
uff, intellij merging apparently keeps the file empty even though I chose "keep the version on master ..."

fm3 reviewed (commented):
Backend looking pretty good already :) I’m wondering if we should move this to multiUser…

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
fix schema, use multi user for keyboard shortcuts and not user

MichaelBuessemeyer committed:
apply backend feedback

MichaelBuessemeyer committed:
fix keyboard shortcuts update query

MichaelBuessemeyer reviewed (commented):
The backend feedback should be applied now. Please give it a small re-check :D

MichaelBuessemeyer committed:
format backend

fm3 commented on app/models/user/MultiUser.scala:
```diff
@@ -207,3 +207,21 @@ class MultiUserDAO @Inject()(sqlClient: SqlClient)(implicit ec: ExecutionContext
+          .as[String])
+    } yield rows.headOption.flatMap(r => JsonHelper.parseAs[JsObject](r).toOption).getOrElse(Json.obj())
+
+  def updateForUser(multiUserId: ObjectId, shortcutsConfig: JsObject): Fox[Unit] =
```
maybe name these *forMultiUser to avoid confusion :)

fm3 reviewed (approved):
Backend LGTM

MichaelBuessemeyer committed:
Apply feedback

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
Merge branch 'master' of github.com:scalableminds/webknossos into custom-keyboard-shortcuts

MichaelBuessemeyer committed:
bump schema version
